Avoid branching on translateUpperBoundForPlutusScripts protocol version

aniketd opened the pull request:
# Description

Closes #5126 

# Checklist

- [x] Commits in meaningful sequence and with useful messages.
- [x] Tests added or updated when needed.
- [x] `CHANGELOG.md` files updated for packages with externally visible changes.  
      **NOTE: _New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd)).
- [ ] ~~Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).~~
- [ ] ~~Version bounds in `.cabal` files updated when necessary.  
      **NOTE: _If bounds change in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process)).~~
- [x] Code formatted (use `scripts/fourmolize.sh`).
- [x] Cabal files formatted (use `scripts/cabal-format.sh`).
- [x] CDDL files are up to date (use `scripts/gen-cddl.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`).
- [x] Self-reviewed the diff.

aniketd committed:
update changelog

Soupstraw commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/TxInfo.hs:
```diff
@@ -218,23 +218,14 @@ transValidityInterval ::
   ValidityInterval ->
   Either a PV1.POSIXTimeRange
-transValidityInterval _ protVer epochInfo systemStart = \case
+transValidityInterval _ epochInfo systemStart = \case
```
I'd rename this to `alonzoTransValidityInterval`, otherwise it's easy to accidentally use this function instead of the one defined in Conway.

lehins replied:
We import TxInfo always qualified. So, I'd prefer not to do this rename

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxInfo.hs:
```diff
@@ -764,3 +769,30 @@ class
   toPlutusChangedParameters _ x = PV3.ChangedParameters (PV3.dataToBuiltinData (toPlutusData x))
+
+-- | Translate a validity interval to POSIX time
+transValidityInterval ::
+  forall proxy era a.
+  Inject (AlonzoContextError era) a =>
+  proxy era ->
+  EpochInfo (Either Text) ->
+  SystemStart ->
+  ValidityInterval ->
+  Either a PV1.POSIXTimeRange
+transValidityInterval _ epochInfo systemStart = \case
```
```suggestion
conwayTransValidityInterval ::
  forall proxy era a.
  Inject (AlonzoContextError era) a =>
  proxy era ->
  EpochInfo (Either Text) ->
  SystemStart ->
  ValidityInterval ->
  Either a PV1.POSIXTimeRange
conwayTransValidityInterval _ epochInfo systemStart = \case
```

lehins replied:
Same here. It would be inconsistent how we deal with TxInfo

Soupstraw commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/TxInfoSpec.hs:
```diff
@@ -54,6 +67,9 @@ spec = do
       expectNoDeposit $ transV10 $ ConwayTxCertDeleg $ ConwayUnRegCert cred SNothing
+
+    it "validity interval's upper bound is open when protocol >= 9" $
+      transVITimeUpperBoundIsOpen @AlonzoEra
```
```suggestion
      transVITimeUpperBoundIsOpen @era
```

teodanciu replied:
I was so confused why this test passes, because at a superficial glance it would seem that it shouldn't, since the exact opposite should happen in Alonzo.  But the era parameter is of course irrelevant, which explains it!

teodanciu commented on eras/conway/impl/CHANGELOG.md:
```diff
@@ -87,6 +89,7 @@
 
 ### `testlib`
 
+* Add `transVITimeUpperBoundIsOpen` test to `TxInfoSpec`.
```
This function doesn't get exported, so don't think there's any point for it to be in the CHANGELOG.

```suggestion
```

teodanciu commented on eras/alonzo/impl/test/Test/Cardano/Ledger/Alonzo/TxInfoSpec.hs:
```diff
@@ -115,8 +115,7 @@ transVITimeUpperBoundIsClosed ::
   Expectation
 transVITimeUpperBoundIsClosed = do
```
As it is, the era parameterization of this function doesn't seem necessary: the function isn't exported, the module is not reused (it's not even part of testlib), and it's invoked only with AlonzoEra, so we might as well just pass AlonzoEra in. 

I was wondering if we could make it a testlib module, then invoke it in subsequent eras - if we generalized it to include the logic of both this and the almost identical one in transVITimeUpperBoundIsOpen (that you moved to Conway), and check the result based on the protocol version.
But I suppose we can't do that, as these are testing two individual functions that are not connected  (not part of a typeclass or so).

Anyway, I would remove the era parameterization, as it doesn't bring anything other than noise here, in my opinion.

aniketd replied:
This is a great idea. Thanks for the suggestion @teodanciu . I will move this into testlib and use in multiple eras in a follow up PR.

teodanciu reviewed (dismissed):
Looks good to me, some minor suggestions and what Joosp said about not invoking the test with @AlonzoEra instead of @era  (though the test still passes and i believe it's technically correct, albeit confusing!)

lehins reviewed (approved):
There is a golden test failure on CI, which @teodanciu can give you a suggestion on how resolve. 
Other than that the PR is good.

teodanciu commented:
> There is a golden test failure on CI, which @teodanciu can give you a suggestion on how resolve.

@aniketd , I think there is is a mistake in the way the golden files were generated! 
For reasons that I cannot remember (but likely as simple as: my stupidity), the `TranslationInstance`s that get serialized in the golden files, use an arbitrary protocol version, here: https://github.com/IntersectMBO/cardano-ledger/blob/master/eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Translation/TranslatableGen.hs#L77.

So I believe what happened is: 
   *  because of this arbitrary protocol version, the  Boolean flag in the translated Interval was set in the golden files to True or False, equally arbitrarily!  
   *  because - until your change -  the same protocol version was used  to check the golden files, the check would always pass. 
   *  now, with your change,  the switching is based on the era rather than on this protocol version, so the checks are failing.
 
That's my theory at least - i got curious because I don't remember seeing these tests fail  before. 
Hope this helps - if what I said doesn't makes sense or you would like to talk about it, let me know please! 

PS:  If my theory is correct, and the golden files are, indeed, wrong - i think it would be good to fix them in a different PR, so that it's clear that your change doesn't alter them  -  so basically to not regenerate the files AND make changes that could potentially change the files in the same PR.

aniketd commented:
@teodanciu Thanks a lot for your investigation into this matter. It helped a great deal in confirming my solution. Please take a second look at this one, now that I have addressed the review comments :grin: and let me know what you think :+1:

aniketd commented:
@aniketd I need to look into the conformance tests and also split the commit into one that changes the code and another that updates the golden files.

teodanciu commented:
Hey @aniketd , FIY: I had a similar failure on my PR, and the good news is that @neilmayhew's [PR](https://github.com/IntersectMBO/cardano-ledger/pull/5248)  has solved mine, so perhaps it will help you too!

aniketd committed:
Make transValidityInterval depend on eras.

...instead of protocol versions.

Implement transValidityInterval specific to Conway.

Remove:
- hardforkConwayTranslateUpperBoundForPlutusScripts
- protocol version argument to transValidityInterval

Move transVITimeUpperBoundIsOpen test to Conway.TxInfoSpec

aniketd committed:
Update golden files

lehins reviewed (approved):
Looks good. Both @teodanciu and @Soupstraw rightfully so pointed out inferior setup for the test that check this functionality. Albeit enough to convince that changes in this PR are sound, but they can be adjusted to make their quality much better. 
I've created #5249, which @aniketd will tackle as a subsequent PR
